### ISSUE TITLE
CachedDataset improvements

### ISSUE THREAD

**Issue by cpcloud** [blaze/blaze#1244]
cc @kwmsmith 

Blaze currently has a `CachedDataset` that can be used to cache the result of expressions. Here's an example:

```python
In [5]: from blaze import Data, CachedDataset, compute, resource, odo

In [6]: import pandas as pd

In [7]: d = Data(CachedDataset(resource('postgresql://localhost/tpc')))

In [8]: d
Out[8]:
Data:       <blaze.cached.CachedDataset object at 0x116d7f0f0>
DataShape:  {
  customer: var * {
    id: int32,
    name: string[25],
    address: string[40],
    nation_id: map[int64, {
      id: int32,
  ...

In [9]: %time d.customer.account_balance.mean()
CPU times: user 1.56 ms, sys: 23 µs, total: 1.58 ms
Wall time: 1.59 ms
Out[9]: 4495.5123316

In [10]: %time d.customer.account_balance.mean()
CPU times: user 32 µs, sys: 1e+03 ns, total: 33 µs
Wall time: 37 µs
Out[10]: 4495.5123316

In [11]: d.data.cache
Out[11]: {mean(_1.customer.account_balance): Decimal('4495.5123316000000000')}
```

This is effective for simple things, but caches everything, and doesn't cache expression subtrees. [`cachey`](https://github.com/blaze/cachey) provides a way to cache analytic computations by scoring values based on the time it takes to compute something and its size. This would be a very welcome addition to `CachedDataset` and would be extremely effective in speeding up computations on things that are very expensive to do IO on such as large CSV files.

One thing to be aware of is the context sensitive nature of certain expressions.

For example, the expression `mean(t.numbers)` is different when its inside of a `By` expression versus not.

```python
avg = t.numbers.mean()
result = by(t.grouper, avg=t.numbers.mean())  # we can't result avg here
```

**Comment by mrocklin** [comment]
Dask's use of cachey lives here and may be an interesting refernce:
https://github.com/blaze/dask/blob/master/dask/cache.py

**Comment by cpcloud** [comment]
Another potential route is to cache everything and bake this into `compute`. That might be hard to extend in the future though.

**Comment by cpcloud** [comment]
Another way--instead of limiting this to `CachedDataset` instances--is to make `cache` a keyword argument to `compute`, like this

```python
import cachey


cache = cachey.Cache()
compute(expr, data, cache=cache)
```

**Comment by llllllllll** [comment]
One thing about the caching of subexprs: We would need to coerce data from the caching format back into the original backend. For example, if we are using postgres, and we cache some subexpr in memory client side, we cannot use that unless we want to odo that into the db somewhere.

**Comment by cpcloud** [comment]
Good point. I wonder if there's a way that we can take advantage of temporary tables here. We'd have to make cache a backend specific thing. I'm not completely sure how this would look. Thoughts welcome here.
